src/Microsoft.ML.TimeSeries/PredictionEngine.cs

```diff
@@ -0,0 +1,253 @@
+                    bool isActive = input.IsColumnActive(c);
+                    differentActive |= wantsActive != isActive;
+
+                    if (wantsActive && !isActive)
+                        throw Contracts.ExceptParam(nameof(input), $"Mapper required column '{input.Schema.GetColumnName(c)}' active but it was not.");
+                }
+
+                outRow = input;
+            }
+
+            // For each of the inner mappers, we will be calling their GetRow method, but to do so we need to know
+            // what we need from them. The last one will just have the input, but the rest will need to be
+            // computed based on the dependencies of the next one in the chain.
+            var deps = new Func<int, bool>[innerMappers.Length];
+            deps[deps.Length - 1] = active;
+            for (int i = deps.Length - 1; i >= 1; --i)
+                deps[i - 1] = innerMappers[i].GetDependencies(deps[i]);
+
+            IRow result = input;
+            for (int i = 0; i < innerMappers.Length; ++i)
+            {
+                Action localDisp;
+                if (innerMappers[i] is CompositeRowToRowMapper)
+                    GetStatefulRows(result, innerMappers[i], deps[i], rows, out localDisp, out result);
+                else
+                    result = innerMappers[i].GetRow(result, deps[i], out localDisp);
+
+                if (result is IStatefulRow)
+                    rows.Add((IStatefulRow)input);
+
+                if (localDisp != null)
+                {
+                    if (disposer == null)
+                        disposer = localDisp;
+                    else
+                        disposer = localDisp + disposer;
+                    // We want the last disposer to be called first, so the order of the addition here is important.
+                }
+            }
+
+            outRow = input;
+        }
+
+        private PredictionEngine(IHostEnvironment env, Func<Schema, IRowToRowMapper> makeMapper, bool ignoreMissingColumns,
+                 SchemaDefinition inputSchemaDefinition, SchemaDefinition outputSchemaDefinition)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            env.AssertValue(makeMapper);
+
+            _inputRow = DataViewConstructionUtils.CreateInputRow<TSrc>(env, inputSchemaDefinition);
+            var mapper = makeMapper(_inputRow.Schema);
+
+            List<IStatefulRow> rows = new List<IStatefulRow>();
+            if (mapper is CompositeRowToRowMapper)
+                GetStatefulRows(_inputRow, mapper, col => true, rows, out _disposer, out var outRow);
+
+            var cursorable = TypedCursorable<TDst>.Create(env, new EmptyDataView(env, mapper.Schema), ignoreMissingColumns, outputSchemaDefinition);
+            var outputRow = mapper.GetRow(_inputRow, col => true, out _disposer);
+
+            if (rows.Count == 0 && outputRow is IStatefulRow)
+                rows.Add((IStatefulRow)outputRow);
+
```

Comment:
this feels like an unroll of 1 level of recursion. Can't `GetStatefulRows` also handle case where mapper is not a `CompositeRowToRowMapper`?